src/docker-extract.c

```diff
@@ -0,0 +1,370 @@
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <fcntl.h>
+#include <sys/file.h>
+#include <sys/param.h>
+#include <sys/stat.h>
+#include <locale.h>
+#include <archive.h>
+#include <archive_entry.h>
+
+#include "config.h"
+#include "util/file.h"
+#include "util/message.h"
+#include "util/registry.h"
+#include "util/util.h"
+
+/* apply_opaque
+ *  Given opq_marker as a path to a whiteout opaque marker
+ *    e.g. usr/share/doc/test/.wh..wh..opq
+ *  Make the directory containing the make opaque for this layer by removing it
+ *  if it exists under rootfs_dir
+ */
+int apply_opaque(const char *opq_marker, char *rootfs_dir) {
+    int retval = 0;
+    char *token;
+    static char buf[MAXPATHLEN];
+
+    token = strrchr(opq_marker, '/');
+    if (token == NULL) {
+        singularity_message(ERROR, "Error getting dirname for opaque marker\n");
+        ABORT(255);
+    }
+
+    retval = snprintf(buf, sizeof(buf), "%s/%s", rootfs_dir, opq_marker);
+    if (retval == -1 || retval >= sizeof(buf)) {
+        singularity_message(ERROR, "Error with pathname too long\n");
+        ABORT(255);
+    }
+
+    // Target may not exist - that's ok
+    retval = 0;
+
+    if (is_dir(buf) == 0) {
+        retval = s_rmdir(buf);
+    }
+
+    return retval;
+}
+
+/* apply_whiteout
+ *  Given wh_marker as a path to a whiteout marker
+ *    e.g. usr/share/doc/test/.wh.deletedfile
+ *  Whiteout the referenced file for this layer by removing it if it exists
+ *  under rootfs_dir
+ */
+int apply_whiteout(const char *wh_marker, char *rootfs_dir) {
+    int retval = 0;
+    char* token;
+    size_t token_pos, l = 0;
+    char buf[MAXPATHLEN];
+
+    token = strstr(wh_marker, ".wh.");
+    if (token == NULL) {
+        singularity_message(ERROR, "Error getting filename for whiteout marker\n");
+        ABORT(255);
+    }
+
+    // Start with ROOTFS
+    retval = snprintf(buf, sizeof(buf), "%s/", rootfs_dir);
+    if (retval == -1 || retval >= sizeof(buf)) {
+        singularity_message(ERROR, "Error with pathname too long\n");
+        ABORT(255);
+    }
+    l = strlen(buf);
+    // Add whiteout path up to .wh.
+    token_pos = strlen(wh_marker) - strlen(token);
+    retval = snprintf(buf + l, token_pos + 1, "%s", wh_marker);
+    if (retval == -1 || retval >= sizeof(buf) - l) {
+        singularity_message(ERROR, "Error with pathname too long\n");
+        ABORT(255);
+    }
+    l = strlen(buf);
+    // Concatenate suffix after .wh
+    retval = snprintf(buf + l, sizeof(buf) - l, "%s", token + 4);
+    if (retval == -1 || retval >= sizeof(buf) - l) {
+        singularity_message(ERROR, "Error with pathname too long\n");
+        ABORT(255);
+    }
+
+    // Target may not exist - that's ok
+    retval = 0;
+
+    if (is_dir(buf) == 0) {
+        retval = s_rmdir(buf);
+    } else if (is_file(buf) == 0) {
```

Comment:
@dctrud symlinks seems not concerned by this check but need to be removed too

Reply:
@cclerget - thanks for spotting that. Added an `|| is_link(buf) == 0` on the else if